### PR TITLE
chore(deps): bump prost to 0.10.1 and tonic to 0.7.1

### LINKED ISSUE #12340: Avoid using the vendored `protoc` from `prost-build`
When #12331 lands, Vector will no longer use the `protoc` command from the `prost-build` crate to compile the necessary protobuf sources. That crate has the option of skipping using the vendored compiler by setting the `PROTO_NO_VENDOR` environment variable. We should further investigate if we can avoid building the vendored code completely by removing the `vendored` feature flag from that crate.

### PR DESCRIPTION
**Pull request by bruceg**
This is a resubmit of #12222 and #12331, along with a resolution for #12340

It turns out the motivation for the previous revert (#12350) was caused by a bind/named configuration issue (#12355) and not a bug in this upgrade as was previously feared. We'll watch the test results carefully on this one to make sure that is the case.

Closes #12340

### CODE REVIEW

**Review by jszwedko — COMMENTED** [review]
Looks like it didn't use the vendored protoc for the `Miscellaneous` check 🤔

**Comment by bruceg** [comment]
I just noticed that pulsar still has a dependency on prost-build v0.8.0. But no, the build error is in 0.10.1. Should the environment variable go in `.bashrc` instead of `.bash_profile`? Is there a better place to put it?

**Comment by jszwedko** [comment]
> I just noticed that pulsar still has a dependency on prost-build v0.8.0. But no, the build error is in 0.10.1. Should the environment variable go in `.bashrc` instead of `.bash_profile`? Is there a better place to put it?

I think you could just add it the `env` block at the top of the Github workflows like: https://github.com/vectordotdev/vector/blob/8287b09672cacd7f921fd6f8c2c5474b5f7f2ac8/.github/workflows/test.yml#L9

It is weird to me that `prost-build` works like that with "autodiscovery". I prefer the `vendored` feature flag approach other crates use.

**Comment by jszwedko** [comment]
I'm not sure if `bashrc` or `bash_profile` is loaded by GHA since it runs commands in "non-interactive" mode.

**Comment by bruceg** [comment]
Okay, but other parts of the bootstraps set `$PATH` in `.bash_profile`, so it _must_ be loaded at some point…

**Comment by bruceg** [comment]
Ahhh, there's a typo! The [in-source documentation (exported at docs.rs) says `$PROTO_NO_VENDOR`](https://docs.rs/prost-build/latest/prost_build/#sourcing-protoc), but [it's actually `$PROTOC_NO_VENDOR`](https://github.com/tokio-rs/prost/blob/8667afcd5ba32afe1a0469c28c9cd8ea8a41fc15/prost-build/build.rs#L91). Will fix.

**Comment by jszwedko** [comment]
> Okay, but other parts of the bootstraps set `$PATH` in `.bash_profile`, so it _must_ be loaded at some point…

Hmm, they do, but:

https://github.com/vectordotdev/vector/blob/8287b09672cacd7f921fd6f8c2c5474b5f7f2ac8/scripts/environment/bootstrap-ubuntu-20.04.sh#L82-L88

Only does it for non-CI.

https://github.com/vectordotdev/vector/blob/8287b09672cacd7f921fd6f8c2c5474b5f7f2ac8/scripts/environment/bootstrap-macos-10.sh#L10-L14

After the `bash_profile` modification you can see it also has special logic for CI to add it to `GITHUB_PATH`.

Those were the only mentions of `bash_profile` I spotted.

**Comment by bruceg** [comment]
> I think you could just add it the `env` block at the top of the Github workflows like:

That was my first thought, actually, but I was concerned I would end up having to put it in a lot of blocks (one for every way we build vector). I didn't investigate how many that would end up being, though.

> It is weird to me that `prost-build` works like that with "autodiscovery". I prefer the `vendored` feature flag approach other crates use.

Agreed. They have _both_ the `vendored` feature flag _and_ autodiscovery. Very odd.

**Comment by jszwedko** [comment]
Yeah, that's fair. You could add it to the bootstrap scripts, I think you'll also just need to add it to `GITHUB_ENV` for CI to pick it up.

File: scripts/environment/bootstrap-macos-10.sh [hunk 1]
```diff
@@ -3,12 +3,15 @@ set -e -o verbose
...
 
-brew install ruby@2.7 coreutils cue-lang/tap/cue
+brew install ruby@2.7 coreutils cue-lang/tap/cue protobuf
 
 gem install bundler
 
 echo "export PATH=\"/usr/local/opt/ruby/bin:\$PATH\"" >> "$HOME/.bash_profile"
 
+# Force the proto-build crate to avoid building the vendored protoc.
+echo "export PROTO_NO_VENDOR=1" >> "$HOME/.bash_profile"
```
Comment (jszwedko) [inline scripts/environment/bootstrap-macos-10.sh:13]: I think you need a similar `echo PROTOC_NO_VENDOR=1 >> "${GITHUB_ENV}"` for this bootstrap script.

**Review by tobz — APPROVED** [review]

**Pull request merged**